### PR TITLE
Title as a block

### PR DESCRIPTION
**Pull request by ellatrix**
This is more of a proof of concept at the moment. There are still some issues to be solved.

Benefits:

* The title becomes optional and filled as you need it.
* The title does not have to be at the top. Some people may see that as a bad thing, but I think this is actually kind of cool. People can be more creative in their layout. E.g. the can add a short intro or something before the title, or maybe a future block that contains post meta. In that way it goes more into the realm of the Customizer.
* Permalink moves to the sidebar, as you should be able to edit it without having a title.
* It becomes part of the table of contents. I event added the error class when other headings are added before the title.
* We can use an editable component and surface formatting in titles. Currently you'd have to write the HTML in the textarea.
* MarkDown shortcut!
* Can be added to the block switcher the user can change some text or a heading to the title.
* Maybe we should even integrate it with the heading block? `[ Title | H2 | H3 | H4 ]`?

Issues:

* We need to convert current titles, and they shouldn't pop up twice for non-Gutenberg themes.

A middle way could be fixing it at the top (for now), but I like the idea of it being a full block.

### CODE REVIEW

**Comment by iseulde** [comment]
Other point: a cover image should be able to act as the title?

**Comment by iseulde** [comment]
Maybe the title even just becomes a value for listing the document somewhere, much like the except, which may or may not be part of the post/document.

**Comment by aduth** [comment]
>The title does not have to be at the top. Some people may see that as a bad thing, but I think this is actually kind of cool. People can be more creative in their layout. E.g. the can add a short intro or something before the title, or maybe a future block that contains post meta. In that way it goes more into the realm of the Customizer.

Wouldn't most themes not be able to support these layouts, since `the_title` and `the_content` are distinct template tags?

**Comment by melchoyce** [comment]
Love the idea of having a cover image take the place of a title!

**Comment by iseulde** [comment]
@aduth Yeah, that's what I meant with 

> We need to convert current titles, and they shouldn't pop up twice for non-Gutenberg themes.

The idea is that `the_title` etc. would go away.

**Comment by westonruter** [comment]
Should not the content of this block get `source`'d from `title` post field instead of the `content` field?

**Comment by StaggerLeee** [comment]
I could almost swear by my left hand yesterday I saw Link option for Title block. Giving me option to edit Link (Slug, or whatever. Only link option, no other toolbar icons.).

And now I cannot reproduce it, after refresh no more. Some glitch, bug, cannot be sure now. It is Demo Post, I did now add anything to it.

PS: I see now, it is a feature not a glitch. Glitch is that I cannot invoke it in Gutenberg Demo Post. Unless I open it via Posts - Gutenberg menu item. Ok, Beta phase, so not so important now.

File: blocks/library/title/index.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from 'i18n';
```
Comment (aduth) [inline blocks/library/title/index.js]: Note with #2172, these need to be updated to prefix the dependencies with `@wordpress/`. You will need to perform a rebase against the latest version of master and apply your changes:

```
git fetch origin
git rebase origin/master
```

File: editor/sidebar/post-permalink/index.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from 'i18n';
```
Comment (aduth) [inline editor/sidebar/post-permalink/index.js]: Note with #2172, these need to be updated to prefix the dependencies with `@wordpress/`. You will need to perform a rebase against the latest version of master and apply your changes:

```
git fetch origin
git rebase origin/master
```

**Comment by youknowriad** [comment]
@iseulde Do you think we should close this and revisit as part of the Template Editor in the customization focus?

**Comment by iseulde** [comment]
I don't mind closing for now, but I do think it would be interesting to have a movable title.

**Comment by danielbachhuber** [comment]
👍 I think this is an interesting idea too, although it runs into some re-world implementation challenges.

**Pull request closed without merging**